### DISCUSSION TITLE
FluentTheme: some TemplateBinding properties should be changed via template parent in control theme.

### DISCUSSION THREAD

**Discussion by rabbitism** [AvaloniaUI/Avalonia#9657] [Ideas]
Raise a discussion first and consider move to issue later. 
Right now, it is very unintuitive to restyle some controls in several pseudoclasses. 
Take below as example. When developer wants to override pointerover background of an ListBoxItem
```xaml
  <ControlTheme x:Key="{x:Type ListBoxItem}" TargetType="ListBoxItem">
    <Setter Property="Background" Value="Transparent" />
    <Setter Property="Template">
      <ControlTemplate>
        <ContentPresenter Name="PART_ContentPresenter"
                          Background="{TemplateBinding Background}"/>
      </ControlTemplate>
    </Setter>

    <!--  PointerOver State  -->
    <Style Selector="^:pointerover /template/ ContentPresenter#PART_ContentPresenter">
      <Setter Property="Background" Value="{DynamicResource SystemControlHighlightListLowBrush}" />
    </Style>
  </ControlTheme>
```
Developers would expect to override by 
```xaml
  <ControlTheme x:Key="NewTheme" BaseOn="{x:Type ListBoxItem}" TargetType="ListBoxItem">
    <Style Selector="^:pointerover">
      <Setter Property="Background" Value="Red" />
    </Style>
  </ControlTheme>
```

But it is not working, because pointerover style is directly applied on ContentPresenter in template.
However, ContentPresenter background comes from template parent, and it is the only consumer of this property. In such a case there is no reason to apply the style directly to template part. It should be applied to the control.

**Comment by jp2masa** [comment]
#2662

**Reply by robloo** [reply]
That discussion is out of date now due to LocalValue priority in styles changing to template and the new addition of ControlTheme. The restrictions encountered no longer seem to be as much of a problem.

**Reply by rabbitism** [reply]
> That discussion is out of date now due to LocalValue priority in styles changing to template and the new addition of ControlTheme. The restrictions encountered no longer seem to be as much of a problem.

Correct, I have tried my own template for several controls and it works as expected in this way. That's why I think it should be back ported to fluent.

**Comment by robloo** [comment]
I would support this in ControlTheme. It does logically make sense and I think the previous limitations preventing it no longer apply.

We need buy in from the core team though. It's also a lot of control themes that need updating.

**Comment by grokys** [comment]
Not 100% sure what you mean here, but I think the crux of what you're saying is:

> The `:pointerover` background color should be applied to the `ListBoxItem` and not the `ListBoxItem`'s `ContentPresenter`

Is that correct?

The problem with that is twofold:

- `ListBoxItem.Background` should not be observed to change when the pointer is over the control
  - See WPF/UWP/HTML to observe that this doesn't happen in other UI frameworks either
- If this were done, and the `ListBoxItem.Background` property were set with a local value (e.g. `ListBoxItem Background="Red"` in the XAML) then the `ListBoxItem` background would no longer change color on pointer-over as the style value has lower priority

But maybe I'm misunderstanding, because the discussion about no longer using `LocalValue` priority in template values don't really apply to this I don't think?

**Reply by robloo** [reply]
> ListBoxItem.Background should not be observed to change when the pointer is over the control
>>See WPF/UWP/HTML to observe that this doesn't happen in other UI frameworks either

I had forgotten about that but it does sound correct. That alone probably stops this idea from moving forward (again).

> But maybe I'm misunderstanding, because the discussion about no longer using LocalValue priority in template values don't really apply to this I don't think?

No, you implemented that change. If there is any misunderstanding it certainly is on my part.
